File: leetcode-js/199-binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    'use strict';

    if (!root) {
        return [];
    }

    const result = [];
    let queue = [root];

    while (queue.length) {
        const len = queue.length;
        for (let i = 0; i < len; i += 1) {
            const node = queue.shift();
            if (i === 0) {
                result.push(node.val);
            }
            if (node.right) {
                queue.push(node.right);
            }
            if (node.left) {
                queue.push(node.left);
            }
        }
    }

    return result;
};

File: leetcode-java/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new LinkedList<Integer>();
        helper(res, root, 0);
        return res;
    }

    void helper(List<Integer> res, TreeNode node, int depth) {
        if (node == null) {
            return;
        }

        if (depth == res.size()) {
            res.add(node.val);
        }

        helper(res, node.right, depth + 1);
        helper(res, node.left, depth + 1);
    }
}
